File: app/src/main/java/com/example/android/timemanagementapp/ui/activities/MainActivity.java
package com.example.android.timemanagementapp.ui.activities;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.content.res.Configuration;
import android.graphics.Color;
import android.os.Bundle;
import android.view.MenuItem;

import com.example.android.timemanagementapp.R;
import com.example.android.timemanagementapp.databinding.ActivityMainBinding;
import com.example.android.timemanagementapp.models.Task;
import com.example.android.timemanagementapp.persistence.TaskRepository;
import com.example.android.timemanagementapp.ui.adapters.TaskAdapter;
import com.google.android.material.navigation.NavigationView;

import java.util.ArrayList;
import java.util.List;


public class MainActivity extends AppCompatActivity implements
        TaskAdapter.OnTaskClickListener,
        NavigationView.OnNavigationItemSelectedListener {

    private ActivityMainBinding binding;

    private ActionBarDrawerToggle drawerToggle; //triple horizontal bar
    private TaskAdapter mAdapter;

    private List<Task> mTasks = new ArrayList<>();
    private TaskRepository mTaskRepository;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        binding.navigationView.setNavigationItemSelectedListener(this);
        binding.navigationView.setCheckedItem(R.id.menu_inbox);

        setSupportActionBar(binding.mainToolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);
        getSupportActionBar().setTitle("Inbox");

        drawerToggle = new ActionBarDrawerToggle(this, binding.drawerLayout, R.string.open, R.string.close);
        binding.drawerLayout.addDrawerListener(drawerToggle);

        mAdapter = new TaskAdapter(MainActivity.this, mTasks, this);
        binding.recyclerViewTasks.setLayoutManager(new LinearLayoutManager(this));
        binding.recyclerViewTasks.setHasFixedSize(true);
        new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT) {
            @Override
            public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
                Task taskToDelete = mTasks.get(viewHolder.getAdapterPosition());

                mTasks.remove(taskToDelete);
                mAdapter.notifyDataSetChanged();
                mTaskRepository.deleteTask(taskToDelete);
            }
        }).attachToRecyclerView(binding.recyclerViewTasks);
        binding.recyclerViewTasks.setAdapter(mAdapter);

        binding.addTaskBtn.setColorFilter(Color.WHITE);
        binding.addTaskBtn.setOnClickListener(view -> {
            Intent intent = new Intent(MainActivity.this, EditTaskActivity.class);
            startActivity(intent);
        });

        mTaskRepository = new TaskRepository(this);
        retrieveTasks();
    }

    private void retrieveTasks() {
        mTaskRepository.retrieveTasks().observe(this, tasks -> {
            mTasks.clear();
            if (tasks != null) {
                mTasks.addAll(tasks);
            }
            mAdapter.notifyDataSetChanged();
        });
    }

    // --- Methods for Drawer Layout ---

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        drawerToggle.syncState();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        drawerToggle.onConfigurationChanged(newConfig);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (drawerToggle.onOptionsItemSelected(item)) return true;
        return super.onOptionsItemSelected(item);
    }

    // --- Methods for Navigation ---

    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_priority:
                getSupportActionBar().setTitle("Priority");
                binding.drawerLayout.closeDrawers();
                mTaskRepository.retrievePriorityTasks().observe(this, tasks -> {
                    mTasks.clear();
                    if (tasks != null) {
                        mTasks.addAll(tasks);
                    }
                    mAdapter.notifyDataSetChanged();
                });
                return true;
            case R.id.menu_inbox:
                getSupportActionBar().setTitle("Inbox");
                binding.drawerLayout.closeDrawers();
                mTaskRepository.retrieveTasks().observe(this, tasks -> {
                    mTasks.clear();
                    if (tasks != null) {
                        mTasks.addAll(tasks);
                    }
                    mAdapter.notifyDataSetChanged();
                });
                return true;
            case R.id.menu_completed:
                getSupportActionBar().setTitle("Completed");
                binding.drawerLayout.closeDrawers();
                mTaskRepository.retrieveCompletedTasks().observe(this, tasks -> {
                    mTasks.clear();
                    if (tasks != null) {
                        mTasks.addAll(tasks);
                    }
                    mAdapter.notifyDataSetChanged();
                });
                return true;
        }
        return false;
    }

    // --- Methods for TaskAdapter.OnTaskClickListener ---
    @Override
    public void onClick(int position) {
        Task task = mTasks.get(position);
        Intent intent = new Intent(MainActivity.this, EditTaskActivity.class);
        intent.putExtra("selected_task", task);
        startActivity(intent);
    }
}

File: app/src/main/java/com/example/android/timemanagementapp/ui/activities/MainActivity.kt
package com.example.android.timemanagementapp.ui.activities

import android.content.Intent
import android.content.res.Configuration
import android.graphics.Color
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.android.timemanagementapp.R
import com.example.android.timemanagementapp.models.Task
import com.example.android.timemanagementapp.persistence.TaskRepository
import com.example.android.timemanagementapp.ui.adapters.TaskAdapter
import com.google.android.material.navigation.NavigationView
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener,
        TaskAdapter.OnTaskClickListener {

    private lateinit var drawerToggle: ActionBarDrawerToggle

    private lateinit var mTaskAdapter: TaskAdapter
    private lateinit var mTaskRepository: TaskRepository
    private var mTasks = ArrayList<Task>()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        navigationView.setNavigationItemSelectedListener(this)
        navigationView.setCheckedItem(R.id.menu_inbox)

        setSupportActionBar(mainToolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeButtonEnabled(true)
        supportActionBar?.title = "Inbox"

        drawerToggle = ActionBarDrawerToggle(this, drawerLayout,R.string.open, R.string.close)
        drawerLayout.addDrawerListener(drawerToggle)

        recyclerTasks.layoutManager = LinearLayoutManager(this)
        recyclerTasks.setHasFixedSize(true)
        mTaskAdapter = TaskAdapter(this, mTasks, this)
        recyclerTasks.adapter = mTaskAdapter
        ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT) {

            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                var taskToDelete: Task = mTasks[viewHolder.adapterPosition]
                // TODO: why do I delete it from tasks and adapter ??
                mTasks.remove(taskToDelete)
                mTaskAdapter.notifyDataSetChanged()
                mTaskRepository.deleteTask(taskToDelete)
            }

        }).attachToRecyclerView(recyclerTasks)

        addTaskBtn.setColorFilter(Color.WHITE)
        addTaskBtn.setOnClickListener { view ->
            val intent = Intent(this, TaskActivity::class.java)
            startActivity(intent)
        }

        mTaskRepository = TaskRepository(this)
        retrieveTasks()
    }

    private fun retrieveTasks() {
        mTaskRepository.retrieveTasks().observe(this, Observer { tasks: List<Task>? ->
            mTasks.clear()
            if (tasks != null) {
                mTasks.addAll(tasks)
            }
            mTaskAdapter.notifyDataSetChanged()
        })
    }

    // --- Methods for Drawer Layout ---
    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        drawerToggle.syncState()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        drawerToggle.onConfigurationChanged(newConfig)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (drawerToggle.onOptionsItemSelected(item)) return true
        return super.onOptionsItemSelected(item)
    }

    // --- Methods for NavigationView.OnNavigationItemSelectedListener ---
    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_priority -> {
                supportActionBar!!.title = "Priority"
                drawerLayout.closeDrawers()
                mTaskRepository.retrievePriorityTasks().observe(this, Observer { tasks: List<Task>? ->
                    mTasks.clear()
                    if (tasks != null) {
                        mTasks.addAll(tasks)
                    }
                    mTaskAdapter.notifyDataSetChanged()
                })
                return true
            }
            R.id.menu_inbox -> {
                supportActionBar!!.title = "Inbox"
                drawerLayout.closeDrawers()
                mTaskRepository.retrieveTasks().observe(this, Observer { tasks: List<Task>? ->
                    mTasks.clear()
                    if (tasks != null) {
                        mTasks.addAll(tasks)
                    }
                    mTaskAdapter.notifyDataSetChanged()
                })
                return true
            }
            R.id.menu_completed -> {
                supportActionBar!!.title = "Completed"
                drawerLayout.closeDrawers()
                mTaskRepository.retrieveCompletedTasks().observe(this, Observer { tasks: List<Task>? ->
                    mTasks.clear()
                    if (tasks != null) {
                        mTasks.addAll(tasks)
                    }
                    mTaskAdapter.notifyDataSetChanged()
                })
                return true
            }
        }
        return false
    }

    // --- Methods for TaskAdapter.OnTaskClickListener ---
    override fun onClick(position: Int) {
        var task: Task = mTasks[position]
        var intent = Intent(this, TaskActivity::class.java)
        intent.putExtra("selected_task", task)
        startActivity(intent)
    }
}